test/SynTreeSpec.hs

```diff
@@ -35,15 +35,15 @@ validBoundsSynTree = do
   maxConsecutiveNegations <- choose (0, 3)
   availableAtoms <- sublistOf ['A' .. 'Z'] `suchThat` (not . null)
   minAmountOfUniqueAtoms <- choose (1, fromIntegral $ length availableAtoms)
-  minNodes <- choose (minAmountOfUniqueAtoms * 2, 60) `suchThat` \minNodes' -> maxConsecutiveNegations /= 0 || odd minNodes'
+  minNodes <- choose (max 3 (minAmountOfUniqueAtoms * 2), 60) `suchThat` \minNodes' -> maxConsecutiveNegations /= 0 || odd minNodes'
   let minDepth = 1 + floor (logBase (2 :: Double) $ fromIntegral minNodes)
   let minMaxDepth = max (maxConsecutiveNegations + 1) minDepth
   maxDepth <- choose (minMaxDepth,max (minMaxDepth+ 3) 10)
   maxNodes <- choose (minNodes, maxNodesForDepth maxDepth) `suchThat`
     \maxNodes' -> (maxConsecutiveNegations /= 0 || odd maxNodes')
       && maxDepth <= maxDepthForNodes maxConsecutiveNegations maxNodes'
   let availableBinOpsCount = if allowArrowOperators then fromIntegral $ length [minBound .. maxBound :: BinOp] else 2
-  minUniqueBinOperators <- choose (1, min (minDepth - 1) availableBinOpsCount)
+  minUniqueBinOperators <- choose (1, min (minDepth - 1) (min availableBinOpsCount ((minNodes `div` 2) - 1)))
```

Comment:
Nested `min`s are probably nicer rewritten using `minimum` on a list.

Also, the current generator here now would "outlaw" `minNodes = 7` and `minUniqueBinOperators = 3`, even though that is a setting that in principle should be okay.

Reply:
I adjusted it.